### PR TITLE
Bonnie measure loader

### PR DESCRIPTION
**Pull request by daco101**
This PR adds measure loading functionality (adapted from bonnie_bundler), but directly to the new measure model.
This PR also adds the batch vsac support.

Pull requests into cqm-parsers require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/BONNIE-1473
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass

**Reviewer 1:**

Name: @jbradl11 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Reviewer 2:**

Name: @hossenlopp 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -2,17 +2,27 @@ source 'https://rubygems.org'
...
 group :development, :test do
   gem 'bundler-audit'
   gem 'rubocop', '~> 0.52.1', require: false
+  gem 'byebug', '~> 6.0.2',  platforms: [:ruby_20, :ruby_21, :ruby_22, :ruby_23]
+  gem 'pry'
+  gem 'pry-nav'
 end
 
 group :development do
+  gem 'clipboard'
```
Comment (hossenlopp) [inline Gemfile]: This doesn't seem like a secure gem to have installed.

File: test/vcr_setup.rb [hunk 2]
```diff
@@ -16,4 +16,20 @@
   c.filter_sensitive_data('<VSAC_USERNAME>') {ENV['VSAC_USERNAME']}
   c.filter_sensitive_data('<VSAC_PASSWORD>') {CGI.escape(ENV['VSAC_PASSWORD'])}
   c.default_cassette_options = { record: :once }
+
+  # Add a custom matcher for use with the bulk request by typheous, so we can ignore service ticket
+  c.register_request_matcher :uri_no_st do |req1, req2|
+    # binding.pry
```
Comment (jbradl11) [inline test/vcr_setup.rb]: binding.pry

File: lib/hqmf-parser/2.0/document.rb [hunk 3]
```diff
@@ -192,7 +192,8 @@ def read_attribute(attribute)
       value_obj = handle_attribute_value(attribute, value) if attribute.at_xpath('./cda:value', NAMESPACES)
 
       # Handle the cms_id - changed to eCQM in MAT 5.4 (QDM 5.3)
-      @cms_id = "CMS#{value}v#{@hqmf_version_number.to_i}" if (name.include? 'eMeasure Identifier') || (name.include? 'eCQM Identifier')
+      # TODO: what is this
```
Comment (jbradl11) [inline lib/hqmf-parser/2.0/document.rb]: TODO comment needs updating

File: Gemfile (hunk 1 above)
Comment (daco101) [reply Gemfile]: How so? The functionality or some other aspect?

Comment (hossenlopp) [reply Gemfile]: It provides access to the system clipboard. What was the reasoning for its addition?

Comment (daco101) [reply Gemfile]: To make it easy to copy a big blob of text to the clipboard during debugging. For example, something like Clipboard.copy(something.to_json), and then pasting that somewhere for easy examination. It's not necessary and I can remove it if needed.

File: lib/measure-loader/cql_loader.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      measure_files.cql_libraries.push(*component_cql_library_files)
+      measure_files.cql_libraries.uniq! { |cla| cla.id + cla.version }
+      return nil
+    end
+
+    # Creates and returns a measure 
+    def create_measure(measure_files)
+      hqmf_model = HQMF::Parser::V2CQLParser.new.parse(measure_files.hqmf_xml)
+
+      measure_files.cql_libraries.each { |cql_lib_files| modify_elm_vs_stuff(cql_lib_files.elm) }
```
Comment (jbradl11) [inline lib/measure-loader/cql_loader.rb]: This line should have a comment. I recommend also changing the function name from using `stuff` to something more descriptive

File: lib/measure-loader/mat_measure_files.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            elsif doc.root.name == 'library'
+              measure_files[:elm_xmls] << doc
+            end
+          end
+        end
+        raise MeasureLoadingException.new("Measure folder found with no hqmf") if measure_files[:hqmf_xml].nil?
+        return measure_files
+      end
+
+      def get_library_version(cql, elm, elm_annotation)
```
Comment (jbradl11) [inline lib/measure-loader/mat_measure_files.rb]: I am not sure what the purpose of this function is, let's talk this morning.

Comment (jbradl11) [reply lib/measure-loader/mat_measure_files.rb]: Rename function to "verify" or "get"

File: lib/measure-loader/elm_dependency_finder.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            deps[parent_name] = [] unless deps.key?('parent_name')
+          end
+          obj.each_pair do |k,v|
+            make_statement_deps_for_elm_helper(v, parent_name, deps, lib_alias_to_path) unless k == 'annotation'
+          end
+        end
+      end
+      
+      def deep_add_external_library_deps(statement, needed_deps_map, all_elms_dep_map)
+        s_library = statement[:library_name]
```
Comment (jbradl11) [inline lib/measure-loader/elm_dependency_finder.rb]: Please rename `s_` to `statement_`

File: lib/measure-loader/mat_measure_files.rb [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        @elm_xml = elm_xml
+      end
+    end
+
+    def initialize(hqmf_xml, cql_libraries, human_readable)
+      @hqmf_xml = hqmf_xml
+      @cql_libraries = cql_libraries
+      @human_readable = human_readable
+
+      raise MeasureLoadingException.new("Measure missing all components.") if @hqmf_xml.nil? || @human_readable.nil? || @cql_libraries.nil? || @cql_libraries.empty?
```
Comment (jbradl11) [inline lib/measure-loader/mat_measure_files.rb]: I would modify the error message, since "components" have a special meaning (component measures).

File: lib/measure-loader/mat_measure_files.rb [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    end
+
+    def self.create_from_zip_file(zip_file)
+      folders = unzip_measure_zip_into_hash(zip_file).values
+      raise MeasureLoadingException.new("No measure found") if folders.empty?
+      folders.sort_by! { |h| h[:depth] }
+      raise MeasureLoadingException.new("Multiple measure folders at top level") if folders[0][:depth] == folders.dig(1,:depth)
+      
+      measure_folder, *composite_measure_folders = folders
```
Comment (jbradl11) [inline lib/measure-loader/mat_measure_files.rb]: `*component_measure_folders`

File: lib/measure-loader/mat_measure_files.rb [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            folders[pn.dirname][:depth] =  pn.each_filename.count
+          end
+        end
+        return folders
+      end
+
+      def make_measure_artifacts(measure_files)
+        library_count = measure_files[:cqls].length
+        unless (measure_files[:elm_xmls].length == library_count && measure_files[:elms].length == library_count)
+          raise MeasureLoadingException.new("Each library must have a cql file, an eml file, and an elm annotation file.")
```
Comment (jbradl11) [inline lib/measure-loader/mat_measure_files.rb]: Change to capitalize "CQL", and list the the "ELM JSON and ELM XML file"

File: lib/measure-loader/elm_dependency_finder.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        needed_deps_map[main_cql_library_id] = all_elms_dep_map[main_cql_library_id]
+        needed_deps_map[main_cql_library_id].each_value do |stmnts|
+          stmnts.each { |stmnt| deep_add_external_library_deps(stmnt, needed_deps_map, all_elms_dep_map) }
+        end
+        return needed_deps_map
+      end
+
+      private
+
+      def elm_id(elm)
```
Comment (jbradl11) [inline lib/measure-loader/elm_dependency_finder.rb]: Is it worth pulling this out into a helper file?

File: lib/measure-loader/value_set_helpers.rb [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        # Once found set the Data Criteria's 'code_list_id' to our fake oid. Do the same for source data criteria.
+        hqmf_model_hash[:data_criteria].each do |data_criteria_name, data_criteria|
+          next unless data_criteria[:inline_code_list] && !data_criteria[:code_list_id]
+          # Check to see if inline_code_list contains the correct code_system and code for a direct reference code.
+          data_criteria[:inline_code_list].each do |code_system, code_list|
+            # Loop over all single code reference objects.
+            value_sets_from_single_code_references.each do |value_set|
+              # If Data Criteria contains a matching code system, check if the correct code exists in the data critera values.
+              # If both values match, set the Data Criteria's 'code_list_id' to the single_code_object_guid.
+              raise StandardError "Expected a singleton" unless value_set.concepts.length == 1
```
Comment (jbradl11) [inline lib/measure-loader/value_set_helpers.rb]: I think this error message should be more descriptive.

File: lib/measure-loader/value_set_helpers.rb [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+            value_set['profile'] = URI.decode(value_set['version'].split('urn:hl7:profile:').last)
+            value_set['version'] = nil
+          # If value has a version and it starts with 'urn:hl7:version:' then strip that and keep the actual version value.
+          elsif value_set['version']&.include?('urn:hl7:version:')
+            value_set['version'] = URI.decode(value_set['version'].split('urn:hl7:version:').last)
+          end
+        end
+      end
+
+      # Removes 'urn:oid:' from ELM for Bonnie and Parse the JSON
```
Comment (jbradl11) [inline lib/measure-loader/value_set_helpers.rb]: Update comment (remove the parse the JSON)

File: lib/measure-loader/value_set_helpers.rb [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      end
+      
+      private
+
+      def get_all_codes_and_code_names(value_sets)
+        all_codes_and_code_names = {}
+        value_sets.each do |value_set|
+          code_sets = {}
+          value_set.concepts.each do |concept|
+            # //TODO: move away from code system names?? not sure if relevant here
```
Comment (jbradl11) [inline lib/measure-loader/value_set_helpers.rb]: Update TODO comment if still needed.

File: lib/measure-loader/value_set_loader.rb [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+          needed_value_sets << {value_set:  value_set, 
+                                vs_vsac_options: vs_vsac_options,
+                                query_version: query_version,
+                                cache_key: cache_key}
+        end
+      end
+
+      vs_responses = @api.get_multiple_valuesets(needed_value_sets)
+
+      [needed_value_sets,vs_responses].transpose.each do |needed_vs,vs_data|
```
Comment (jbradl11) [inline lib/measure-loader/value_set_loader.rb]: Interesting!

File: lib/measure-loader/value_set_loader.rb [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        return { version: value_set[:version] } unless value_set[:version].nil?
+      end
+      return @vsac_options
+    end
+
+    def modelize_value_set(vsac_xml_response, query_version)
+      doc = Nokogiri::XML(vsac_xml_response)
+      doc.root.add_namespace_definition("vs","urn:ihe:iti:svs:2008")
+      vs_element = doc.at_xpath("/vs:RetrieveValueSetResponse/vs:ValueSet|/vs:RetrieveMultipleValueSetsResponse/vs:DescribedValueSet")
+      raise error "BLAH" if vs_element.nil?
```
Comment (jbradl11) [inline lib/measure-loader/value_set_loader.rb]: Error message needs updating.

File: lib/measure-loader/cql_loader.rb [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+      component_measures = []
+      if measure_files.components.present?
+        measure, component_measures = create_measure_and_components(measure_files)
+      else
+        measure = create_measure(measure_files)
+      end
+
+      measure.package = CQM::MeasurePackage.new(file: BSON::Binary.new(measure_zip.read))
+      measures = component_measures << measure
```
Comment (jbradl11) [inline lib/measure-loader/cql_loader.rb]: I don't really have a suggestion here, this just looks a bit odd. Maybe pre-define `measures`? We can discuss tomorrow.

File: Gemfile (hunk 1 above)
Comment (jbradl11) [reply Gemfile]: I would recommend removing this for now.

File: lib/util/vsac_api.rb [hunk 17]
```diff
@@ -262,6 +263,19 @@ def get_valueset(oid, options = {})
         end
       end
 
+      def get_multiple_valuesets(needed_value_sets)
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: Add a comment matching how other functions are commented in this file.

File: lib/util/vsac_api.rb [hunk 18]
```diff
@@ -287,6 +301,90 @@ def get_ticket_granting_ticket(username, password)
...
+
+        raise Util::VSAC::VSNotFoundError.new(expected_oid) unless (vs_element && vs_element['ID'] == expected_oid)
+        raise Util::VSAC::VSEmptyError.new(expected_oid) if concepts.empty?
+        return vs_data
+      end
+
+      def get_multiple_valueset_raw_responses(needed_value_sets)
+        service_tickets = get_service_tickets(needed_value_sets.size)
+
+        hydra = Typhoeus::Hydra.new
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: Since this is new functionality using a new library, I recommend adding a comment what this is doing.

File: lib/util/vsac_api.rb [hunk 19]
```diff
@@ -287,6 +301,90 @@ def get_ticket_granting_ticket(username, password)
...
+          hydra.queue(request)
+          request
+        end
+
+        hydra.run
+        responses = requests.map(&:response)
+        return responses
+      end
+
+      def get_service_tickets(amount)
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: Comments needed and this seems somewhat duplicating functionality in `getTicket`

File: lib/util/vsac_api.rb [hunk 20]
```diff
@@ -287,6 +301,90 @@ def get_ticket_granting_ticket(username, password)
...
+        end
+        return tickets
+      end
+
+      def create_valueset_request(oid, ticket, options = {})
+        # base parameter oid is always needed
+        params = { id: oid }
+
+        # release parameter, should be used moving forward
+        params[:release] = options[:release] unless options[:release].nil?
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: Some duplication of code between here and `get_valueset`

**Comment by jbradl11** [comment]
I recommend having a test for comparing CQL statement dependencies the before and after.

**Comment by jbradl11** [comment]
Also, please add a section to the readme on how to use the Measure parser and explain what you pass in and what it passes back to you with a little example.
